components/dashboard/src/teams/sso/OIDCClientConfigModal.tsx

```diff
@@ -0,0 +1,146 @@
+    onClose: () => void;
+};
+
+export const OIDCClientConfigModal: FC<Props> = ({ clientConfig, onClose }) => {
+    const { data: org } = useCurrentOrg();
+    const upsertClientConfig = useUpsertOIDCClientMutation();
+
+    const isNew = !clientConfig;
+
+    const [issuer, setIssuer] = useState(clientConfig?.oidcConfig?.issuer ?? "");
+    const [clientId, setClientId] = useState(clientConfig?.oauth2Config?.clientId ?? "");
+    const [clientSecret, setClientSecret] = useState(clientConfig?.oauth2Config?.clientSecret ?? "");
+
+    const redirectUrl = gitpodHostUrl.with({ pathname: `/iam/oidc/callback` }).toString();
+
+    const issuerError = useOnBlurError(`Please enter a valid URL.`, issuer.trim().length > 0 && isURL(issuer));
+    const clientIdError = useOnBlurError("Client ID is missing.", clientId.trim().length > 0);
+    const clientSecretError = useOnBlurError("Client Secret is missing.", clientSecret.trim().length > 0);
+
+    const isValid = useMemo(
+        () => [issuerError, clientIdError, clientSecretError].every((e) => e.isValid),
+        [clientIdError, clientSecretError, issuerError],
+    );
+
+    const saveConfig = useCallback(async () => {
+        if (!org) {
+            console.error("no current org selected");
+            return;
+        }
+        if (!isValid) {
+            return;
+        }
+
+        const trimmedIssuer = issuer.trim();
+        const trimmedClientId = clientId.trim();
+        const trimmedClientSecret = clientSecret.trim();
+
+        try {
+            await upsertClientConfig.mutateAsync({
+                config: isNew
+                    ? {
+                          organizationId: org.id,
+                          oauth2Config: {
+                              clientId: trimmedClientId,
+                              clientSecret: trimmedClientSecret,
+                          },
+                          oidcConfig: {
+                              issuer: trimmedIssuer,
+                          },
+                      }
+                    : {
+                          id: clientConfig?.id,
+                          organizationId: org.id,
+                          oauth2Config: {
+                              clientId: trimmedClientSecret,
+                              // TODO: determine how we should handle when user doesn't change their secret
+                              clientSecret: clientSecret === "redacted" ? "" : trimmedClientSecret,
+                          },
+                          oidcConfig: {
+                              issuer: trimmedIssuer,
+                          },
+                      },
+            });
+
+            onClose();
+        } catch (error) {
+            console.error(error);
+        }
+    }, [clientConfig?.id, clientId, clientSecret, isNew, isValid, issuer, onClose, org, upsertClientConfig]);
+
+    const errorMessage = upsertClientConfig.isError ? "There was a problem saving your configuration." : "";
```

Comment:
It would be great to surface those errors.

Especially for the backend validation those errors are crucial to identify connectivity problems between Gitpod's backend and the IdP services. The client side validation is nice and a good improvement, but if the backend cannot connect to the services, the auth processes will break too late on receiving the callback from IdP.

For example if you enter some valid URL, which might point to non-existing host, we see the following error from backend. I agree it's a bit too long, we might need to tweak that (on backend) as well. While this is about non exiting host, same applies for hosts which are not reachable from the network the backend services are running in, so basically it's a whole class of connectivity issues.

> Get "https://google234444444.com": dial tcp: lookup google234444444.com on 10.43.0.10:53: no such host

Reply:
Yah, I don't have the modal-based error message component built in a way that supports a dynamic amount of content very well atm, but I agree, we should be able to include those details in the error message, just need to make it more robust so it can expand to fit content, but also not block the modal body content it slides over. I'll tackle that in a follow-up.